openedx_tagging/core/tagging/api.py

```diff
@@ -136,3 +136,19 @@ def tag_object(
     Preserves existing (valid) tags, adds new (valid) tags, and removes omitted (or invalid) tags.
     """
     return taxonomy.cast().tag_object(tags, object_id)
+
+
+def autocomplete_tags(
+    taxonomy: Taxonomy,
+    search: str,
+    object_id: str = None,
+) -> QuerySet:
+    """
+    Returns tag values that contains the given search string.
+    The result is returned in alphabetical order.
+
+    The output is a QuerySet of dictionaries in with `_value` and `tag`.
+
+    Subclasses can override this method to perform their own autocomplete process.
```

Comment:
I think we can say more here about the purpose of this api function.

And note: subclasses cannot override this method because it's an API, not a model method.

```suggestion
    Provides auto-complete suggestions by matching the `search` string against existing object tags linked to the given taxonomy. A case-insensitive search is used in order to return the highest number of relevant tags.
    
    If `object_id` is provided, then object tag values already linked to this object are omitted from the returned suggestions. (Object tag values must be unique for a given object + taxonomy, and so omitting these suggestions helps users avoid duplication errors.)
    
    Returns a QuerySet of dictionaries containing distinct `value` (string) and `tag` (numeric ID) values, sorted alphabetically by `value`. The `value` is what should be shown as a suggestion to users, and if it's a free-text taxonomy, `tag` will be `None`:  we include the `tag` ID in anticipation of the second use case listed below.
    
    Use cases:
    * This method is useful for reducing tag variation in free-text taxonomies by showing users tags that are similar to what they're typing. E.g., if the `search` string "dn" shows that other objects have been tagged with "DNA", "DNA electrophoresis", and "DNA fingerprinting", this encourages users to use those existing tags if relevant, instead of creating new ones that look similar (e.g. "dna finger-printing").
    * It could also be used to assist tagging for closed taxonomies with a list of possible tags which is too large to return all at once, e.g. a user model taxonomy that dynamically creates tags on request for any registered user in the database. (Note that this is not implemented yet, but may be as part of a future change.)
```

Reply:
Updated here: https://github.com/openedx/openedx-learning/pull/59/commits/593c047b8d7bf5c3a314b8a25d19f55beb5238c8